### PR TITLE
New connector for working with the Catalist Match API

### PR DESCRIPTION
**Pull request by austinweisgrau**
# Connector for working with the Catalist Match API.

The Catalist Match tool requires OAuth2.0 client credentials for the API as well as
credentials for accessing the Catalist sftp bucket. Each Catalist client is given
their own bucket alias named after a tree species, used for constructing the
filepath within the sftp bucket.

Accessing the Catalist sftp bucket and Match API both require the source IP address
to be explicitly white-listed by Catalist.

Example usage:
```
tbl = Table.from_csv(...)
client = CatalistMatch(...)
match_result = client.match(tbl)
```

# Tests
Tests are included in this PR.

# Changes to parsons
Two changes are made to the parsons package outside of this new connector.

28949a9a20fe4422aae3b0ac6373e912c61d4b4c makes a small change to the api_connector error handler to look for error messages in a response `text` attribute, where the error message is found in responses from the Catalist API.

1ffb04631bc79e99bbd523bdf314ba179849104a adds a development dependency `pytest-mock` to support the ergonomic use of mocking in pytest. This dependency is then used in the tests on the catalist match connector in 31f3a1816cdb09d5004fb11f70d1bd6da0182cd1.

# Expected future changes
Following a community development session, JSON responses from the API will be parsed using python dataclasses. Currently JSON responses are returned raw, which matches the normal pattern across parsons connectors.

### CODE REVIEW

**Comment by austinweisgrau** [comment]
Made some updates to type hints to be backwards compatible with Python versions back to 3.6

**Comment by austinweisgrau** [comment]
@IanRFerguson or @shaunagm Do you know what's up with that docker failure? When I hit details I just get `{"error": "Our service is temporarily unavailable. We'll be back soon"}`

**Comment by austinweisgrau** [comment]
> @IanRFerguson or @shaunagm Do you know what's up with that docker failure? When I hit details I just get `{"error": "Our service is temporarily unavailable. We'll be back soon"}`

NVM I got it with [001b527](https://github.com/move-coop/parsons/pull/912/commits/001b527477d9096ba064adfb9c9a419b405c89e0)

File: parsons/catalist/catalist.py
Comment (IanRFerguson) [inline parsons/catalist/catalist.py:23]: Makes sense! Do we anticipate Catalist making other templates available? If so might make sense to paramaterize this in the API call with the current value as a default, but I'll defer to you here

File: parsons/catalist/catalist.py
Comment (IanRFerguson) [inline parsons/catalist/catalist.py:29]: Maybe in the docs we can offer a broad overview of how to obtain client credentials from Catalist / what other setup steps are required

File: parsons/catalist/catalist.py
Comment (IanRFerguson) [inline parsons/catalist/catalist.py:98]: Can we get `Args:` in this docstring?

File: parsons/catalist/catalist.py
Comment (IanRFerguson) [inline parsons/catalist/catalist.py:245]: Missing args in docstring

File: parsons/catalist/catalist.py
Comment (IanRFerguson) [inline parsons/catalist/catalist.py:298]: Love this, very elegant solution

File: parsons/catalist/catalist.py
Comment (IanRFerguson) [inline parsons/catalist/catalist.py:327]: Is there any extra level of detail we can provide here as to why the job failed? Even including the error code from the response might be useful

File: parsons/catalist/catalist.py
Comment (IanRFerguson) [inline parsons/catalist/catalist.py:383]: Love this logic

**Review by IanRFerguson — CHANGES_REQUESTED** [review]
Looks **excellent** overall - you crushed this! I have a few comments throughout but nothing structural or functional, the code is super legible and easy to navigate.

One note - I think adding a `docs/catalist.rst` file is required before merging. I left a comment on setup / obtaining client creds and I reckon the docs would be a good place to outline that

File: parsons/catalist/catalist.py
Comment (austinweisgrau) [reply parsons/catalist/catalist.py:265]: Done with 561deccce2f6a219c0270c382c2906805ecdbd75

File: parsons/catalist/catalist.py
Comment (austinweisgrau) [reply parsons/catalist/catalist.py]: Done with 561deccce2f6a219c0270c382c2906805ecdbd75

File: parsons/catalist/catalist.py
Comment (austinweisgrau) [reply parsons/catalist/catalist.py:29]: Asking Catalist for this

File: parsons/catalist/catalist.py
Comment (austinweisgrau) [reply parsons/catalist/catalist.py:363]: Asking Catalist for this

Comment (austinweisgrau) [reply parsons/catalist/catalist.py]: Done with a37527a99c2ace60e26e30dc9cc59bfc141a3967

File: parsons/catalist/catalist.py
Comment (austinweisgrau) [reply parsons/catalist/catalist.py]: Done with 08b48339a9721afe29c115f7b2fa47650ac3d852

File: parsons/catalist/catalist.py
Comment (austinweisgrau) [reply parsons/catalist/catalist.py:30]: Slightly implemented with 74cd6c2924a9363b094065370780ece01c980b6c and I'm asking if they can provide any more info here

**Comment by austinweisgrau** [comment]
> Looks **excellent** overall - you crushed this! I have a few comments throughout but nothing structural or functional, the code is super legible and easy to navigate.
> 
> One note - I think adding a `docs/catalist.rst` file is required before merging. I left a comment on setup / obtaining client creds and I reckon the docs would be a good place to outline that

Added the `docs/catalist.rst` file with 9b735233c193be56535e14fec0c137a97ab88a37, that was new for me though so not totally sure I did that correctly.

**Review by IanRFerguson — APPROVED** [review]
Thanks for addressing the comments so thoroughly! Looks great

File: requirements.txt
Comment (shaunagm) [inline requirements.txt:50]: Is this used in the tests? I see putest and requests-mock being imported but not pytest-mock

Comment (austinweisgrau) [reply requirements.txt:50]: It creates a pytest fixture called "mocker", which is used in the pytest fixtures in[ this file](https://github.com/austinweisgrau/parsons/blob/match/test/test_catalist/conftest.py)

**Pull request merged**
